src/AWS/Lambda/Events/Kafka.hs

```diff
@@ -0,0 +1,249 @@
+well as self-managed Kafka clusters.
+
+Lambda considers Amazon Managed Streaming for Kafka (MSK) to be a
+different event source type from a self-managed Apache Kafka cluster,
+but their payloads are very similar. The types in this module are
+derived from inspecting invocation payloads, and from reading the
+following links:
+
+  * https://docs.aws.amazon.com/lambda/latest/dg/with-kafka.html
+  * https://docs.aws.amazon.com/lambda/latest/dg/with-msk.html
+  * https://aws.amazon.com/blogs/compute/using-amazon-msk-as-an-event-source-for-aws-lambda/
+  * https://aws.amazon.com/blogs/compute/using-self-hosted-apache-kafka-as-an-event-source-for-aws-lambda/
+
+-}
+
+module AWS.Lambda.Events.Kafka (
+  KafkaEvent(..),
+  EventSource(..),
+  Record,
+  Record'(..),
+  Header (..),
+  Timestamp(..),
+  -- * Internal
+  parseTimestamp,
+  unparseTimestamp,
+  int64ToUTCTime,
+  utcTimeToInt64
+) where
+
+import           Data.Aeson
+import           Data.Aeson.Types
+import           Data.ByteString        (ByteString)
+import qualified Data.ByteString        as B
+import qualified Data.ByteString.Base64 as B64
+import           Data.List.NonEmpty     (NonEmpty)
+import qualified Data.List.NonEmpty     as NE
+import           Data.Function          ((&))
+import           Data.Int               (Int32, Int64)
+import           Data.Map               (Map)
+import           Data.Maybe             (catMaybes, maybeToList)
+import           Data.Scientific        (toBoundedInteger)
+import           Data.Semigroup         ((<>))
+import           Data.Text              (Text)
+import qualified Data.Text              as T
+import qualified Data.Text.Encoding     as TE
+import           Data.Time              (UTCTime)
+import           Data.Time.Clock.POSIX  (posixSecondsToUTCTime,
+                                         utcTimeToPOSIXSeconds)
+import           Data.Vector            (Vector)
+import           GHC.Generics           (Generic)
+
+#if MIN_VERSION_aeson(2,0,0)
+import qualified Data.Aeson.Key         as Key
+import qualified Data.Aeson.KeyMap      as KM
+
+keyToText :: Key -> Text
+keyToText = Key.toText
+
+keyFromText :: Text -> Key
+keyFromText = Key.fromText
+#else
+import qualified Data.HashMap.Strict    as KM
+type Key = Text
+
+keyToText :: Key -> Text
+keyToText = id
+
+keyFromText :: Text -> Key
+keyFromText = id
+#endif
+
+-- | Represents an event from either Amazon MSK or a self-managed
+-- Apache Kafka cluster, as the payloads are very similar.
+--
+-- The 'ToJSON' and 'FromJSON' instances on 'Record' perform base64
+-- conversion for you.
+--
+-- See the <https://docs.aws.amazon.com/lambda/latest/dg/with-kafka.html AWS documentation>
+-- for a sample payload.
+data KafkaEvent = KafkaEvent {
+  eventSource      :: !EventSource,
+  -- | Only present when using the "Amazon MSK" event source mapping.
+  eventSourceArn   :: !(Maybe Text),
+  bootstrapServers :: !(NonEmpty Text),
```

Comment:
I need to double-check that this field is present when using the MSK subscription type.

Reply:
If there's any uncertainty, it might be safer to simply use `[]`.  `NonEmpty` seems like a decent application here if at least one is guaranteed, but I admittedly often find it's machinery clunky to actually use.

Reply:
I'm pretty sure that it does provide this field. If it does not, then `bootstrapServers` would more likely be missing from the payload completely, anyway.